pubsub/pubsub_test.go

```diff
@@ -0,0 +1,94 @@
+package pubsub_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/tendermint/tmlibs/pubsub"
+	"github.com/tendermint/tmlibs/pubsub/query"
+)
+
+func TestSubscribe(t *testing.T) {
+	t.Parallel()
+
+	s := pubsub.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch := s.Subscribe(query.Empty{})
+	s.Publish("Goblin Queen")
+	assertReceive(t, "Goblin Queen", ch)
+}
+
+func TestQueries(t *testing.T) {
+	t.Parallel()
+
+	s := pubsub.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch1 := s.Subscribe(q("tm.events.type=NewBlock"))
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock"})
+	assertReceive(t, "Iceman", ch1)
+
+	ch2 := s.Subscribe(q("tm.events.type=NewBlock AND abci.account.name=Igor"))
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock", "abci.account.name": "Igor"})
+	assertReceive(t, "Iceman", ch2)
+
+	ch3 := s.Subscribe(q("tm.events.type=NewBlock AND abci.account.name=Igor"))
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock"})
+	<-ch1
+	assert.Zero(t, len(ch3))
+}
+
+func TestPublish(t *testing.T) {
+	t.Parallel()
+
+	s := pubsub.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch := s.Subscribe(query.Empty{})
+	s.Publish("Ka-Zar")
```

Comment:
> Otherwise a few poorly implemented (or malicious) clients could halt the tendermint server.

well, yes and no. because this is not the end client. In TM it's all up to us how to implement the end client (WSClient) https://github.com/tendermint/tendermint/blob/ec6341215e61d04892c263318a8d06313c3f6968/rpc/core/events.go#L20. There we just drop the messages if the WebSocket queue is full.